### ISSUE TITLE
App Name not found

### ISSUE THREAD

**Issue by sdwaa** [ptarmiganlabs/butler-sos#272]
**Describe the bug**
cannot get the App nema for Apps table

**To Reproduce**
select * from apps , eg : in_memory_docs_names , active_docs_names

**Expected behavior**
based on my understanding, should ge the exact name of the Apps

**Screenshots**
select  in_memory_docs_names from apps
![image](https://user-images.githubusercontent.com/94944764/143374395-2ebe8e43-d895-495d-a262-d2a60cba0d9a.png)

**Describe environment:**
 - OS: e.g. Windows Server 2016
 - Containerisation : no
 - Version of Butler SOS used : 6.1.2
 - Command used to start Butler SOS : node butler-sos.js

**Config file(s)**
What's the content of the config file(s) you use?
```yaml
---
Butler-SOS:
  # All configuration items are mandatory, unless otherwise noted.

  # Logging configuration
  logLevel: info          # Log level. Possible log levels are silly, debug, verbose, info, warn, error
  fileLogging: true       # true/false to enable/disable logging to disk file
  logDirectory: log       # Subdirectory where log files are stored
  anonTelemetry: true     # Can Butler SOS send anonymous data about what computer it is running on? 
                          # More info on whata data is collected: https://butler-sos.ptarmiganlabs.com/docs/about/telemetry/
                          # Please consider leaving this at true - it really helps future development of Butler SOS!

  # Heartbeats can be used to send "I'm alive" messages to some other tool, e.g. an infrastructure monitoring tool
  # The concept is simple: The remoteURL will be called at the specified frequency. The receiving tool will then know 
  # that Butler SOS is alive.
  heartbeat:
    enable: false
    remoteURL: http://my.monitoring.server/some/path/
    frequency: every 1 hour         # https://bunkat.github.io/later/parsers.html

  # Docker health checks are used when running Butler SOS as a Docker container. 
  # The Docker engine will call the container's health check REST endpoint with a set interval to determine
  # whether the container is alive/well or not.
  # If you are not running Butler SOS in Docker you can safely disable this feature. 
  dockerHealthCheck:
    enable: false                    # Control whether a REST endpoint will be set up to serve Docker health check messages
    port: 12398                     # Port the Docker health check service runs on (if enabled)

  # Uptime monitor
  uptimeMonitor:
    enable: true                    # Should uptime messages be written to the console and log files?
    frequency: every 15 minutes     # https://bunkat.github.io/later/parsers.html
    logLevel: verbose               # Starting at what log level should uptime messages be shown?
    storeInInfluxdb: 
      butlerSOSMemoryUsage: true    # Should data on Butler SOS' own memory use be stored in Infludb?
      instanceTag: POC              # Tag that can be used to differentiate data from multiple Butler SOS instances

  # Track individual users opening/closing apps and starting/stopping sessions. 
  # Requires log appender XML file(s) to be added to Sense server(s.
  userEvents:                       
    enable: false
    excludeUser:                    # Optional blacklist of users that should be disregarded when it comes to user events
      - directory: LAB
        userId: testuser1
      - directory: LAB
        userId: testuser2
    udpServerConfig:
      serverHost: <IP or FQDN>      # Host/IP where user event server will listen for events from Sense
      portUserActivityEvents: 9997  # Port on which user event server will listen for events from Sense
    tags:                           # Tags are added to the data before it's stored in InfluxDB
      - tag: env
        value: DEV
      - tag: foo
        value: bar
    sendToMQTT: 
      enable: false                 # Set to true if user events should be forwarded as MQTT messages
      topic: userEvent              # MQTT topic to use for user event messages
    sendToInfluxdb:
      enable: true                  # Set to true if user events should be stored in InfluxDB

  # Qlik Sense logging db config parameters
  logdb:
    enable: true
    # Items below are mandatory if logdb.enable=true
    pollingInterval: 60000            # How often (milliseconds) should Postgres log db be queried for warnings and errors?
    queryPeriod: 5 minutes            # How far back should Butler SOS query for log entries? Default is 5 min
    host: cnpek013a8.----------
    port: 4432
    qlogsReaderUser: qlogs_reader
    qlogsReaderPwd: -----
    extractErrors: true               # Should error level entries be extracted from log db into Influxdb?
    extractWarnings: true             # Should warn level entries be extracted from log db into Influxdb?
    extractInfo: false                 # Should info level entries be extracted from log db into Influxdb? 
                                      # Warning! Seting this to true will result in LOTS of log messages 
                                      # being retrrieved by Butler SOS!

  # Certificates to use when connecting to Sense. Get these from the Certificate Export in QMC.
  cert:
    rejectUnauthorized: true           # Set to false to ignore warnings/errors caused by Qlik Sense's self-signed certificates.
                                        # Set to true if the Qlik Sense root CA is available on the computer where Butler SOS is running.
    clientCert: D:\butlerSOS\cnpek01968\cnpek01968.pem
    clientCertKey: D:\butlerSOS\cnpek01968\cnpek01968_key.pem
    clientCertCA: D:\butlerSOS\cnpek01968\root.pem
    clientCertPassphrase: <certificate key password, if one was specified when exporting certificates from Sense QMC >
    # If running Butler SOS in a Docker container, the cert paths MUST be the following
    # clientCert: /nodeapp/config/certificate/client.pem
    # clientCertKey: /nodeapp/config/certificate/client_key.pem
    # clientCertCA: /nodeapp/config/certificate/root.pem
    # clientCertPassphrase:

  # MQTT config parameters
  mqttConfig:
    enable: false
    # Items below are mandatory if mqttConfig.enable=true
    brokerHost: <IP of MQTT broker/server>
    brokerPort: 1883
    baseTopic: butler-sos/          # Topic should end with /

  # Prometheus config
  # If enabled, select Butler SOS metrics will be exposed on a Prometheus compatible URL from where they can be scraped.
  prometheus:
    enable: false                                    # Default false
    host: <IP or FQDN where Butler SOS is running>  # On what IP/FQDN should the Prometheus metrics be exposed? Default 0.0.0.0, i.e. all available IPs
    port: 9842                                      # Port for Prometheus endpoint. Default 9842

  # Influx db config parameters
  influxdbConfig:
    enable: true
    # Items below are mandatory if influxdbConfig.enable=true
    hostIP: cnpek0178h
    hostPort: 8086    # Optional. Default value=8086
    auth:
      enable: false                 # Does influxdb instance require authentication (true/false)?
      username: <username>          # Username for Influxdb authentication. Mandatory if auth.enable=true
      password: <password>          # Password for Influxdb authentication. Mandatory if auth.enable=true
    dbName: SenseOps

    # Default retention policy that should be created in InfluxDB when Butler SOS creates a new database there. 
    # Any data older than retention policy threshold will be purged from InfluxDB.
    retentionPolicy:
      name: 10d
      duration: 10d

    # Control whether certain fields are stored in InfluxDB or not
    # Use with caution! Enabling activeDocs, loadedDocs or inMemoryDocs may result in lots of data sent to InfluxDB.
    includeFields:
      activeDocs: false              # Should data on what docs are active be stored in Influxdb (true/false)? 
      loadedDocs: false              # Should data on what docs are loaded be stored in Influxdb (true/false)?
      inMemoryDocs: false            # Should data on what docs are in memory be stored in Influxdb (true/false)?

  # Extract app names
  appNames: 
    enableAppNameExtract: true    # Extract app names in addition to app IDs (tue/false)?
    extractInterval: 60000        # How often (milliseconds) should app names be extracted?
    hostIP: cnpek01968.--------          # What Sense server should be queried for app names?

  # Sessions per virtual proxy
  userSessions:
    enableSessionExtract: true      # Query unique user IDs of what users have sessions open (true/false)?
    # Items below are mandatory if enableSessionExtract=true    
    pollingInterval: 30000        # How often (milliseconds) should session data be polled?
    excludeUser:                  # Optional blacklist of users that should be disregarded when it comes to session monitoring.
                                  # Blacklist is only applied to data in InfluxDB. All session data will be sent to MQTT.
      - directory: LAB
        userId: testuser1
      - directory: LAB
        userId: testuser2

  serversToMonitor:
    pollingInterval: 30000          # How often (milliseconds) should the healthcheck API be polled?

    # If false, Butler SOS will accept TLS certificates on the server without verifying them with the CA. 
    # If true, data will only be retrieved from the Sense server if that server's TLS cert verifies 
    # successfuully against the list of CAs available on the computer where Butler SOS is running.
    rejectUnauthorized: true 

    # List of extra tags for each server. Useful for creating more advanced Grafana dashboards.
    # Each server below MUST include these tags in its serverTags property.
    # The tags below are just examples - define your own as needed
    serverTagsDefinition: 
      - server_group
      - serverLocation
      - server-type
      - serverBrand

    # Sense Servers that should be queried for healthcheck data 
    servers:
      - host: cnpek01968:4747
        serverName: cnpek01968
        serverDescription: Develop Node
        logDbHost: cnpek013a8.------
        userSessions:
          enable: true
          # Items below are mandatory if userSessions.enable=true
          host: cnpek01968:4243
          virtualProxies:
            - virtualProxy: /win                 # Default virtual proxy
            #- virtualProxy: /hdr              # "hdr" virtual proxy
            #- virtualProxy: /sales            # "sales" virtual proxy
        serverTags:
          server_group: Central
          serverLocation: Asia
          server-type: virtual
          serverBrand: Dell
````

**Comment by mountaindude** [comment]
Do you get any errors or warnings in the Butler console/log files?
If the extraction of app names for some reason failed you should get a warning or error in the logs, I would think.

I ran this query on a Sense system here and got app names, so the extract logic as such is working.
The apps without app names are session apps.

```
> use senseops
Using database senseops
> select in_memory_docs, in_memory_docs_names from apps order by time desc limit 10
name: apps
time                in_memory_docs                                                            in_memory_docs_names
----                --------------                                                            --------------------
1637783832110869194
1637783832107038734 78e9a4bf-515f-4400-b73d-385f325e98a0,bf083ad3-7d7e-4c59-867b-2912cc39ccb6 Tickets,Laddboxar
1637783817118320960
1637783817111995361 78e9a4bf-515f-4400-b73d-385f325e98a0                                      Laddboxar
1637783802120756530
1637783802114656932 78e9a4bf-515f-4400-b73d-385f325e98a0,bf083ad3-7d7e-4c59-867b-2912cc39ccb6 Tickets,Laddboxar
1637783787131281328 78e9a4bf-515f-4400-b73d-385f325e98a0,bf083ad3-7d7e-4c59-867b-2912cc39ccb6 Tickets,Laddboxar
1637783787126667044
1637783772117666578
1637783772107212216 78e9a4bf-515f-4400-b73d-385f325e98a0,bf083ad3-7d7e-4c59-867b-2912cc39ccb6 Tickets,Laddboxar
>
```

**Comment by sdwaa** [comment]
> 

Dear mountaindude:

thanks for your quickly reply, but may i know where can I find the error logs?

**Comment by mountaindude** [comment]
The logs are always written to the console. If you started Butler SOS from the command line the console is what you are looking at after starting Butler SOS.  If running in Docker the console is available via the `docker logs ... ` command.

Log files are optional, but it seems you have them enabled, with a log directory of `log` under your main Butler directory.
The log files contain the exact same info as the console, minus the colors.

You can also increase the log level verbosity. 
Setting it to `verbose` will give more detailed output, `debug` even more so.

```yaml
Butler-SOS:
  # All configuration items are mandatory, unless otherwise noted.

  # Logging configuration
  logLevel: info          # Log level. Possible log levels are silly, debug, verbose, info, warn, error
  fileLogging: true       # true/false to enable/disable logging to disk file
  logDirectory: log       # Subdirectory where log files are stored
```

**Comment by sdwaa** [comment]
> The logs are always written to the console. If you started Butler SOS from the command line the console is what you are looking at after starting Butler SOS. If running in Docker the console is available via the `docker logs ... ` command.
> 
> Log files are optional, but it seems you have them enabled, with a log directory of `log` under your main Butler directory. The log files contain the exact same info as the console, minus the colors.
> 
> You can also increase the log level verbosity. Setting it to `verbose` will give more detailed output, `debug` even more so.
> 
> ```yaml
> Butler-SOS:
>   # All configuration items are mandatory, unless otherwise noted.
> 
>   # Logging configuration
>   logLevel: info          # Log level. Possible log levels are silly, debug, verbose, info, warn, error
>   fileLogging: true       # true/false to enable/disable logging to disk file
>   logDirectory: log       # Subdirectory where log files are stored
> ```

dear mountaindude:

here is the latest log with all the same info:

2021-11-26T01:06:03.905Z info: APP NAMES: Done getting app names

or should increase the log level?

**Comment by mountaindude** [comment]
First: 
You have set the query interval for getting app names to 60 seconds. 
This means you should wait at least this long before querying InfluxDB for app names. 
At least *new* app names, there may of course be app names in InfluxDB from previous Butler SOS executions.

Secondly:
If you want to check what's stored in InfluxDB you can use queries like the one below from the InfluxDB command line client (which is usually installed alongside the InfluxDB server). 
This query shows all unique combinations of app names that are stored in InfluxDB. 
I.e. all apps that have simultaneously been loaded into this particular Sense server during the time period for which InfluxDB has data.

```
> use senseops
Using database senseops
> select distinct(in_memory_docs_names) from (select in_memory_docs, in_memory_docs_names from apps)
name: apps
time distinct
---- --------
0    78e9a4bf-515f-4400-b73d-385f325e98a0
0
0    Laddboxar
0    Always failing reload (no delay),Laddboxar
0    Laddboxar,NYC parking tickets
0    Laddboxar,NYC parking tickets,NYT comments
0    Kickstarter,Laddboxar,Meetup.com,NYC parking tickets,NYT comments
0    Kickstarter,Laddboxar,Meetup.com,NYC parking tickets,NYT comments,Synt key
0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments
0    Meetup.com
0    Laddboxar,Meetup.com
0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Test data - Seattle checkouts by title
0    Always failing reload (no delay),Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments
0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Training - Field indexing
0    Training - Field indexing
0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Test data - Seattle checkouts by title,Training - Field indexing
0    Test data - Seattle checkouts by title
0    NYT comments
0    B_1_lan,NYT comments
0    App3,Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Training - Field indexing
0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYT comments,Training - Field indexing
0    Meetup.com,NYT comments
0    Butler 7 Slack debug,Kickstarter,Laddboxar,Meetup.com,NYC parking tickets,NYT comments,Training - Field indexing
0    Laddboxar DEV,Meetup.com,NYT comments
0    Laddboxar,Laddboxar DEV,Operations Monitor
0    Laddboxar,Laddboxar DEV
0    78e9a4bf-515f-4400-b73d-385f325e98a0,f42d6b16-8faf-45ca-a783-59f9da47db6d
0    Meetup.com,Test data - Seattle checkouts by title
0    Laddboxar,bf083ad3-7d7e-4c59-867b-2912cc39ccb6
>
```



Third:
Temporarily changing the log level is a good idea.

Setting it to `verbose` will give you a bit more info in the logs, like this:

```
2021-11-26T09:53:10.080Z verbose: APP NAMES: Event started: Get app names
2021-11-26T09:53:10.279Z verbose: APP NAMES: Number of apps: 214
2021-11-26T09:53:10.282Z info: APP NAMES: Done getting app names
```

That tells me that Butler SOS found 214 apps for which it then got app names.

If you want to see the actual names too you can set log level to `debug`.
This will give you **lots** of info in the logs. But in there somewhere there is an entry that tells you the actual app names (I've removed some of those 214 app names from the log below to make it reasonable in size):

```
2021-11-26T09:55:07.941Z verbose: APP NAMES: Number of apps: 214
2021-11-26T09:55:07.941Z debug: APP NAMES: App list JSON: [{"id":"6c4c1cfb-cbae-402b-845e-3966a6903cf5","name":"License Monitor_25.15.2.0"},{"id":"064c9f04-a0d1-4b8c-9a72-f960725af6a0","name":"License Monitor_25.7.3.0"},{"id":"c1e27bf7-56ae-45b9-87a6-cdfab255b269","name":"Partial load 1"},{"id":"5d7ae888-61cd-4539-97b2-6cf5baaa6f9d","name":"2018 sales targets"},{"id":"7169ec96-892d-42b3-acb8-4bc3845a00fa","name":"__1 synth 1"},{"id":"4eb2288b-e26c-4ae7-acee-f8cb458b532b","name":"License Monitor_26.4.1.0"}"}]
2021-11-26T09:55:07.946Z info: APP NAMES: Done getting app names
```

This allows me to verify that Butler SOS really got the app names. 

If the app names still don't show up in InfluxDB (and later in Grafana dashboards) the problem probably lies in InfluxDB or your Grafana dashboard. Seems strange that would happen though... If Butler SOS can write to InfluxDB without reporting any errors, the app names should be there.

**Comment by sdwaa** [comment]
> First: You have set the query interval for getting app names to 60 seconds. This means you should wait at least this long before querying InfluxDB for app names. At least _new_ app names, there may of course be app names in InfluxDB from previous Butler SOS executions.
> 
> Secondly: If you want to check what's stored in InfluxDB you can use queries like the one below from the InfluxDB command line client (which is usually installed alongside the InfluxDB server). This query shows all unique combinations of app names that are stored in InfluxDB. I.e. all apps that have simultaneously been loaded into this particular Sense server during the time period for which InfluxDB has data.
> 
> ```
> > use senseops
> Using database senseops
> > select distinct(in_memory_docs_names) from (select in_memory_docs, in_memory_docs_names from apps)
> name: apps
> time distinct
> ---- --------
> 0    78e9a4bf-515f-4400-b73d-385f325e98a0
> 0
> 0    Laddboxar
> 0    Always failing reload (no delay),Laddboxar
> 0    Laddboxar,NYC parking tickets
> 0    Laddboxar,NYC parking tickets,NYT comments
> 0    Kickstarter,Laddboxar,Meetup.com,NYC parking tickets,NYT comments
> 0    Kickstarter,Laddboxar,Meetup.com,NYC parking tickets,NYT comments,Synt key
> 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments
> 0    Meetup.com
> 0    Laddboxar,Meetup.com
> 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Test data - Seattle checkouts by title
> 0    Always failing reload (no delay),Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments
> 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Training - Field indexing
> 0    Training - Field indexing
> 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Test data - Seattle checkouts by title,Training - Field indexing
> 0    Test data - Seattle checkouts by title
> 0    NYT comments
> 0    B_1_lan,NYT comments
> 0    App3,Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Training - Field indexing
> 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYT comments,Training - Field indexing
> 0    Meetup.com,NYT comments
> 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Meetup.com,NYC parking tickets,NYT comments,Training - Field indexing
> 0    Laddboxar DEV,Meetup.com,NYT comments
> 0    Laddboxar,Laddboxar DEV,Operations Monitor
> 0    Laddboxar,Laddboxar DEV
> 0    78e9a4bf-515f-4400-b73d-385f325e98a0,f42d6b16-8faf-45ca-a783-59f9da47db6d
> 0    Meetup.com,Test data - Seattle checkouts by title
> 0    Laddboxar,bf083ad3-7d7e-4c59-867b-2912cc39ccb6
> >
> ```
> 
> Third: Temporarily changing the log level is a good idea.
> 
> Setting it to `verbose` will give you a bit more info in the logs, like this:
> 
> ```
> 2021-11-26T09:53:10.080Z verbose: APP NAMES: Event started: Get app names
> 2021-11-26T09:53:10.279Z verbose: APP NAMES: Number of apps: 214
> 2021-11-26T09:53:10.282Z info: APP NAMES: Done getting app names
> ```
> 
> That tells me that Butler SOS found 214 apps for which it then got app names.
> 
> If you want to see the actual names too you can set log level to `debug`. This will give you **lots** of info in the logs. But in there somewhere there is an entry that tells you the actual app names (I've removed some of those 214 app names from the log below to make it reasonable in size):
> 
> ```
> 2021-11-26T09:55:07.941Z verbose: APP NAMES: Number of apps: 214
> 2021-11-26T09:55:07.941Z debug: APP NAMES: App list JSON: [{"id":"6c4c1cfb-cbae-402b-845e-3966a6903cf5","name":"License Monitor_25.15.2.0"},{"id":"064c9f04-a0d1-4b8c-9a72-f960725af6a0","name":"License Monitor_25.7.3.0"},{"id":"c1e27bf7-56ae-45b9-87a6-cdfab255b269","name":"Partial load 1"},{"id":"5d7ae888-61cd-4539-97b2-6cf5baaa6f9d","name":"2018 sales targets"},{"id":"7169ec96-892d-42b3-acb8-4bc3845a00fa","name":"__1 synth 1"},{"id":"4eb2288b-e26c-4ae7-acee-f8cb458b532b","name":"License Monitor_26.4.1.0"}"}]
> 2021-11-26T09:55:07.946Z info: APP NAMES: Done getting app names
> ```
> 
> This allows me to verify that Butler SOS really got the app names.
> 
> If the app names still don't show up in InfluxDB (and later in Grafana dashboards) the problem probably lies in InfluxDB or your Grafana dashboard. Seems strange that would happen though... If Butler SOS can write to InfluxDB without reporting any errors, the app names should be there.

Dear mountaindude:

I follow all the steps you mentioned, while unfortunatelly, i still cannot get the Names of the Apps as below:

select distinct(in_memory_docs_names) from (select in_memory_docs, in_memory_docs_names from apps)
name: apps
time distinct
---- --------
0


while on one of the node, i can see the total number of my Apps on the whole platform

2021-11-29T02:15:11.248Z verbose: APP NAMES: Number of apps: 8259
2021-11-29T02:15:11.303Z info: APP NAMES: Done getting app names

after i changed the log level to debug,  I can also get the exact name of my Apps, while still not able to get the names.

**Comment by sdwaa** [comment]
> > First: You have set the query interval for getting app names to 60 seconds. This means you should wait at least this long before querying InfluxDB for app names. At least _new_ app names, there may of course be app names in InfluxDB from previous Butler SOS executions.
> > Secondly: If you want to check what's stored in InfluxDB you can use queries like the one below from the InfluxDB command line client (which is usually installed alongside the InfluxDB server). This query shows all unique combinations of app names that are stored in InfluxDB. I.e. all apps that have simultaneously been loaded into this particular Sense server during the time period for which InfluxDB has data.
> > ```
> > > use senseops
> > Using database senseops
> > > select distinct(in_memory_docs_names) from (select in_memory_docs, in_memory_docs_names from apps)
> > name: apps
> > time distinct
> > ---- --------
> > 0    78e9a4bf-515f-4400-b73d-385f325e98a0
> > 0
> > 0    Laddboxar
> > 0    Always failing reload (no delay),Laddboxar
> > 0    Laddboxar,NYC parking tickets
> > 0    Laddboxar,NYC parking tickets,NYT comments
> > 0    Kickstarter,Laddboxar,Meetup.com,NYC parking tickets,NYT comments
> > 0    Kickstarter,Laddboxar,Meetup.com,NYC parking tickets,NYT comments,Synt key
> > 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments
> > 0    Meetup.com
> > 0    Laddboxar,Meetup.com
> > 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Test data - Seattle checkouts by title
> > 0    Always failing reload (no delay),Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments
> > 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Training - Field indexing
> > 0    Training - Field indexing
> > 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Test data - Seattle checkouts by title,Training - Field indexing
> > 0    Test data - Seattle checkouts by title
> > 0    NYT comments
> > 0    B_1_lan,NYT comments
> > 0    App3,Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYC parking tickets,NYT comments,Training - Field indexing
> > 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Laddboxar DEV,Meetup.com,NYT comments,Training - Field indexing
> > 0    Meetup.com,NYT comments
> > 0    Butler 7 Slack debug,Kickstarter,Laddboxar,Meetup.com,NYC parking tickets,NYT comments,Training - Field indexing
> > 0    Laddboxar DEV,Meetup.com,NYT comments
> > 0    Laddboxar,Laddboxar DEV,Operations Monitor
> > 0    Laddboxar,Laddboxar DEV
> > 0    78e9a4bf-515f-4400-b73d-385f325e98a0,f42d6b16-8faf-45ca-a783-59f9da47db6d
> > 0    Meetup.com,Test data - Seattle checkouts by title
> > 0    Laddboxar,bf083ad3-7d7e-4c59-867b-2912cc39ccb6
> > >
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Third: Temporarily changing the log level is a good idea.
> > Setting it to `verbose` will give you a bit more info in the logs, like this:
> > ```
> > 2021-11-26T09:53:10.080Z verbose: APP NAMES: Event started: Get app names
> > 2021-11-26T09:53:10.279Z verbose: APP NAMES: Number of apps: 214
> > 2021-11-26T09:53:10.282Z info: APP NAMES: Done getting app names
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That tells me that Butler SOS found 214 apps for which it then got app names.
> > If you want to see the actual names too you can set log level to `debug`. This will give you **lots** of info in the logs. But in there somewhere there is an entry that tells you the actual app names (I've removed some of those 214 app names from the log below to make it reasonable in size):
> > ```
> > 2021-11-26T09:55:07.941Z verbose: APP NAMES: Number of apps: 214
> > 2021-11-26T09:55:07.941Z debug: APP NAMES: App list JSON: [{"id":"6c4c1cfb-cbae-402b-845e-3966a6903cf5","name":"License Monitor_25.15.2.0"},{"id":"064c9f04-a0d1-4b8c-9a72-f960725af6a0","name":"License Monitor_25.7.3.0"},{"id":"c1e27bf7-56ae-45b9-87a6-cdfab255b269","name":"Partial load 1"},{"id":"5d7ae888-61cd-4539-97b2-6cf5baaa6f9d","name":"2018 sales targets"},{"id":"7169ec96-892d-42b3-acb8-4bc3845a00fa","name":"__1 synth 1"},{"id":"4eb2288b-e26c-4ae7-acee-f8cb458b532b","name":"License Monitor_26.4.1.0"}"}]
> > 2021-11-26T09:55:07.946Z info: APP NAMES: Done getting app names
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This allows me to verify that Butler SOS really got the app names.
> > If the app names still don't show up in InfluxDB (and later in Grafana dashboards) the problem probably lies in InfluxDB or your Grafana dashboard. Seems strange that would happen though... If Butler SOS can write to InfluxDB without reporting any errors, the app names should be there.
> 
> Dear mountaindude:
> 
> I follow all the steps you mentioned, while unfortunatelly, i still cannot get the Names of the Apps as below:
> 
> select distinct(in_memory_docs_names) from (select in_memory_docs, in_memory_docs_names from apps) name: apps time distinct
> 
> 0
> 
> while on one of the node, i can see the total number of my Apps on the whole platform
> 
> 2021-11-29T02:15:11.248Z verbose: APP NAMES: Number of apps: 8259 2021-11-29T02:15:11.303Z info: APP NAMES: Done getting app names
> 
> after i changed the log level to debug, I can also get the exact name of my Apps, while still not able to get the names.

BTW, if I havee 5 nodes, how many yaml file should I use to configure? currenlty i'm using 5 seperate yaml file and 5 sepereate cmd window to run.

**Comment by mountaindude** [comment]
Long overdue response here, apologies. These are strange and sometimes demanding times.

Butler SOS is designed to run as a single instance, not on each node in a Sense cluster. 
For a small Sense system it is usually ok to run Butler SOS on the Sense server itself, as the overhead created by Butler SOS is small.
For larger systems you probably want to consider running tools like Butler SOS on their own, separate server. I usually call this server MISC or TOOLS. It can be a very small server, a Linux box with a couple of GB RAM works fine.  

But it's a good idea from a systems architecture point of view to keep the Sense servers/nodes as clean as possible.
If nothing else it makes it much easier to replace a Sense node if/when needed.

To summarise: One Butler SOS instance per Sense cluster. If you have separate DEV, TEST and PROD clusters you probably want to have one Butler SOS instance for each of those clusters. 
Several Butler SOS instances can run in parallel on the same server, just make sure there is no overlapping/conflicting configuration between the different instances. For example, each instance should have its own log directory, don't share any network ports etc. 


I am not sure what's going on in your case... Based on the logs you shared: 

```
2021-11-29T02:15:11.248Z verbose: APP NAMES: Number of apps: 8259
2021-11-29T02:15:11.303Z info: APP NAMES: Done getting app names
```

we know that you have a rather large number of apps in your system.
If I am to guess I think we could be looking at a timeout problem... Butler SOS queries the Sense repository database (which is PostgreSQL) for the app names. But if the database doesn't respond in time there might be a timeout.
But then we should have seen a timeout error too... Strange.

We would really have to test connectivity to the repository db, to ensure we can query it as needed.
I'l see if I can whip up a stand-alone little tool that just pulls all app names from repository db, it would be a nice little utility in itself.

**Comment by mountaindude** [comment]
@sdwaa 

I've added a new ticket for creating a small connectivity-checking tool.
https://github.com/ptarmiganlabs/butler-sos/issues/292

This will make it easier to diagnose and understand what's going on in cases like yours.

**Comment by mountaindude** [comment]
@sdwaa 

Another thing came to mind here.
Looking at the log snippet you sent: 

```
2021-11-29T02:15:11.248Z verbose: APP NAMES: Number of apps: 8259
2021-11-29T02:15:11.303Z info: APP NAMES: Done getting app names
```

It looks like you have verbose logging for Butler SOS.
If you change that to debug logging you *should* see a list of all app names, similar to what's shown below in my test Sense environment:

```
2021-11-26T09:55:07.941Z verbose: APP NAMES: Number of apps: 214
2021-11-26T09:55:07.941Z debug: APP NAMES: App list JSON: [{"id":"6c4c1cfb-cbae-402b-845e-3966a6903cf5","name":"License Monitor_25.15.2.0"},{"id":"064c9f04-a0d1-4b8c-9a72-f960725af6a0","name":"License Monitor_25.7.3.0"},{"id":"c1e27bf7-56ae-45b9-87a6-cdfab255b269","name":"Partial load 1"},{"id":"5d7ae888-61cd-4539-97b2-6cf5baaa6f9d","name":"2018 sales targets"},{"id":"7169ec96-892d-42b3-acb8-4bc3845a00fa","name":"__1 synth 1"},{"id":"4eb2288b-e26c-4ae7-acee-f8cb458b532b","name":"License Monitor_26.4.1.0"}"}]
2021-11-26T09:55:07.946Z info: APP NAMES: Done getting app names
```

See that second line saying "debug: ....." ? 
If that line, in your Sense environment shows a long list of app names we know that we in fact get the app names from the repository database. 
The problem is then elsewhere. 

If you *don't* get a long list of app names in that debug output, we know that Butler SOS failed getting app names from repository database.

So, as a first step I suggest you turn on debug logging, start Butler SOS and wait a couple of minutes for the app name extract to do its first run. That will tell us more what's going on behind the scenes.

**Comment by mountaindude** [comment]
Closing due to inactivity.
